### ISSUE TITLE
Buffer overflow in server

### ISSUE THREAD

**Issue by mustakimur** [dianestephens/fileserver_cpp#36]
A buffer overflow has been detected in the server code. In absence of stack protector or stack canary leakage, it can lead to RCE.

Following is the input sequence:
```
connect guest
create user steve
login steve
send file.txt
quit
```

Here is the debug info from server:
```
(gdb) r
Starting program: /home/mustakim/projects/server/src/server 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Ready for connections ...
Connected to 127.0.0.1 on port 51456
[New Thread 0x7ffff7a4f700 (LWP 20707)]
Ready for connections ...
waiting for client cmd
Command: connect guest
login guest
login attempt
usernameeeee> steve
login successful
current user: "steve"
waiting for client cmd
Command: send file.txt
received cmdLine: send file.txt
 size() of cmdLine: 13
fstr from: file.txt to: 13
*** buffer overflow detected ***: terminated

Thread 2 "server" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff7a4f700 (LWP 20707)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7bc9859 in __GI_abort () at abort.c:79
#2  0x00007ffff7c343ee in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7d5e07c "*** %s ***: terminated\n")
    at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007ffff7cd6b4a in __GI___fortify_fail (msg=msg@entry=0x7ffff7d5e012 "buffer overflow detected") at fortify_fail.c:26
#4  0x00007ffff7cd53e6 in __GI___chk_fail () at chk_fail.c:28
#5  0x0000555555573500 in memset (__len=8, __ch=0, __dest=0x7ffff7a4df34) at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:71
#6  connection_thread (p_client_socket=<optimized out>) at server.cpp:439
#7  0x00007ffff7d9f609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#8  0x00007ffff7cc6293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

The vulnerable code is here:
```c++
memset(&filesize, 0, sizeof(&filesize));
recv(clientsocket, &filesize, sizeof(&filesize), 0); // get the size of the file in bytes
```

Although the overflow detected in memset and has low criticality, but the next recv can be used to overflow the same way with user input.

**Comment by irfanOzn** [comment]
Dear Dr. Mustakimur,

Thank You for the bug report. To solve the stack overflow bug here, we replaced 64-bit sizeof(&filesize) length parameter with sizeof(int) to make the receive buffer's length equal to the bytes being read from the socket stream, which is the same technique we used in the client bug. Also, we commented the memset line out. The suggested change is that: (I hope we have correctly fixed that)



![image](https://user-images.githubusercontent.com/76975914/112742273-232ed600-8f5b-11eb-8100-1553fe6d6a9a.png)


Thank You so much,
Team Red.
